File: Programmers/다트 게임.js
// https://programmers.co.kr/learn/courses/30/lessons/17682

function solution(dartResult) {
  var answer = [];

  const result = [];
  let nextFirstIndex = 0;
  let target;

  for (let i = 1; i < dartResult.length; i++) {
    if (!(dartResult[i] === "S" || dartResult[i] === "T" || dartResult[i] === "D")) continue;

    let score;
    let bonus;
    let option;

    if (dartResult[i + 1] === "*" || dartResult[i + 1] === "#") {
      target = dartResult.slice(nextFirstIndex, i + 2).split("");
      nextFirstIndex = i + 2;
      option = target.pop();
    } else {
      target = dartResult.slice(nextFirstIndex, i + 1).split("");
      nextFirstIndex = i + 1;
    }

    bonus = target.pop();
    score = parseInt(target.join(""));

    result.push({ score: score, bonus: bonus, option: option });
  }

  result.forEach((n) => {
    let finalScore = n.score;
    switch (n.bonus) {
      case "D":
        finalScore = Math.pow(finalScore, 2);
        break;
      case "T":
        finalScore = Math.pow(finalScore, 3);
        break;
      default:
        break;
    }
    if (n.option === "#") finalScore *= -1;
    else if (n.option === "*") {
      finalScore *= 2;
      if (answer.length) answer[answer.length - 1] *= 2;
    }
    answer.push(finalScore);
  });

  return answer.reduce((acc, cur) => acc + cur, 0);
}

// const dartResult = "1S2D*3T";
const dartResult = "1D2S#10S";
// const dartResult = "1D2S0T";
console.log(solution(dartResult));

File: Programmers/다트 게임.swift
// https://programmers.co.kr/learn/courses/30/lessons/17682?language=swift
// 다트 게임

import Foundation

func solution(_ dartResult:String) -> Int {
    let powDic: Dictionary<Character, Double> = ["S": 1.0, "D": 2.0, "T": 3.0]
    var scores: [Int] = []
    var startIdx = dartResult.startIndex
    
    for i in 0..<dartResult.count {
        let targetIdx = dartResult.index(dartResult.startIndex, offsetBy: i)
        let target: Character = dartResult[targetIdx]
        if target == "S" || target == "D" || target == "T" {
            var score: Int = Int(String(dartResult[startIdx..<targetIdx])) ?? 0
            score = Int(pow(Double(score), powDic[target]!))
            scores.append(score)
            
            if i == dartResult.count - 1 { break } 
            
            startIdx = dartResult.index(after: targetIdx)
            let next: Character = dartResult[startIdx]
            if next != "*" && next != "#" { continue }
            startIdx = dartResult.index(after: startIdx)
            
            if next == "#" { scores[scores.index(before: scores.endIndex)] *= -1 }
            else {
                scores[scores.index(before: scores.endIndex)] *= 2
                if scores.count > 1 {
                    scores[scores.index(scores.endIndex, offsetBy: -2)] *= 2
                }
            }
        }
    }
    
    return scores.reduce(0, { acc, cur in acc + cur})
}